File: LeetCode/top-interview-questions-medium/Backtracking/generate-parenthesis.py


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        if n > 0:
            s = [""]*2*n
            self._generate_parenthesis(s,0,n,0,0, ans)
            ans.sort()
        return ans

    def _generate_parenthesis(self, s, index, n, _open, _close, ans):
        if _close == n:
            ans.append("".join(s))
        else:
            if _open > _close:
                s[index] = ')'
                self._generate_parenthesis(s, index+1, n, _open, _close+1, ans)
            if _open < n:
                s[index] = '('
                self._generate_parenthesis(s, index+1, n, _open+1, _close, ans)

File: LeetCode/top-interview-questions-medium/Backtracking/generate-parenthesis.cpp
// NOT WORKING


/* Author: Jitendra Kumar (jitendra.iitp@gmail.com) */
#include<bits/stdc++.h>
using namespace std;
#define endl '\n'
#define INF LLONG_MAX/4
#define MOD 1e9+7
#define DEBUG(x) cerr << #x << " is " << (x) << endl;

class Solution {
private:
    void _generate(vector<string> &ans, int N, string &s, int idx, int open, int close) {
        if(open==N && close==N) {
            ans.push_back(s);
            return;
        }
        if(open <= close) {
            s[idx] = '(';
            _generate(ans, N, s, idx+1, open+1, close);
        } else {
            s[idx] = ')';
            _generate(ans, N, s, idx+1, open, close+1);
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string s(2*n, '$');
        int open=0, close=1, idx=0;
        _generate(ans, ans.size(), s, idx, open, close);
        sort(ans.begin(), ans.end());
        return ans;
    }
};

